perception/autoware_ground_segmentation/src/scan_ground_filter/grid_ground_filter.cpp

```diff
@@ -0,0 +1,460 @@
+        // this point is obstacle
+        out_no_ground_indices.indices.push_back(pt_idx);
+      } else if (
+        abs(height) < global_slope_threshold && abs(height) < param_.non_ground_height_threshold) {
+        // this point is ground
+        ground_bin.addPoint(radius, height, pt_idx);
+        is_ground_found = true;
+      }
+    }
+    cell.is_processed_ = true;
+    cell.has_ground_ = is_ground_found;
+    if (is_ground_found) {
+      cell.is_ground_initialized_ = true;
+      ground_bin.processAverage();
+      cell.avg_height_ = ground_bin.getAverageHeight();
+      cell.avg_radius_ = ground_bin.getAverageRadius();
+      cell.max_height_ = ground_bin.getMaxHeight();
+      cell.min_height_ = ground_bin.getMinHeight();
+      cell.gradient_ = std::clamp(
+        cell.avg_height_ / cell.avg_radius_, -param_.global_slope_max_ratio,
+        param_.global_slope_max_ratio);
+      cell.intercept_ = 0.0f;
+    } else {
+      cell.is_ground_initialized_ = false;
+    }
+  }
+}
+
+void GridGroundFilter::SegmentContinuousCell(
+  const Cell & cell, PointsCentroid & ground_bin, pcl::PointIndices & out_no_ground_indices)
+{
+  const Cell & prev_cell = grid_ptr_->getCell(cell.scan_grid_root_idx_);
+  const float local_thresh_angle_ratio = std::tan(DEG2RAD(5.0));
+
+  // loop over points in the cell
+  for (const auto & pt : cell.point_list_) {
+    const size_t & pt_idx = pt.index;
+    const float & radius = pt.distance;
+    const float & height = pt.height;
+
+    // 1. height is out-of-range
+    const float delta_z = height - prev_cell.avg_height_;
+    if (delta_z > param_.detection_range_z_max) {
+      // this point is out-of-range
+      continue;
+    }
+
+    // 2. the angle is exceed the global slope threshold
+    if (height > param_.global_slope_max_ratio * radius) {
+      // this point is obstacle
+      out_no_ground_indices.indices.push_back(pt_idx);
+      // go to the next point
+      continue;
+    }
+
+    // 3. local slope
+    const float delta_radius = radius - prev_cell.avg_radius_;
+    if (abs(delta_z) < param_.global_slope_max_ratio * delta_radius) {
+      // this point is ground
+      ground_bin.addPoint(radius, height, pt_idx);
+      // go to the next point
+      continue;
+    }
+
+    // 3. height from the estimated ground
+    const float next_gnd_z = cell.gradient_ * radius + cell.intercept_;
+    const float gnd_z_local_thresh = local_thresh_angle_ratio * delta_radius;
+    const float delta_gnd_z = height - next_gnd_z;
+    const float gnd_z_threshold = param_.non_ground_height_threshold + gnd_z_local_thresh;
+    if (delta_gnd_z > gnd_z_threshold) {
+      // this point is obstacle
+      out_no_ground_indices.indices.push_back(pt_idx);
+      // go to the next point
+      continue;
+    }
+    if (abs(delta_gnd_z) <= gnd_z_threshold) {
+      // this point is ground
+      ground_bin.addPoint(radius, height, pt_idx);
+      // go to the next point
+      continue;
+    }
```

Comment:
[imho]

Same comment with other segment function.
Beaware underground pointclouds will not categorized into ground nor obstacle.